server/src/main/java/org/candlepin/model/EntitlementCurator.java

```diff
@@ -285,67 +293,102 @@ private Criteria createModifiesDateFilteringCriteria(Consumer consumer, Date sta
-        List<Product> products = productCurator.listAllByIds(entitlement.getOwner(), pidEnts.keySet());
+        List<Product> overlappingProducts =
+                productCurator.listAllByIds(pidEnts.getAllProductIds());
 
-        for (Product p : products) {
-            boolean modifies = p.modifies(entitlement.getPool().getProductId());
-            Iterator<Product> ppit = entitlement.getPool().getProvidedProducts().iterator();
+        Set<String> entitlementProducts = new HashSet<String>();
+        for (Entitlement entitlement : entitlements) {
+            entitlementProducts.add(entitlement.getPool().getProductId());
+            for (Product product : entitlement.getPool().getProvidedProducts()) {
+                entitlementProducts.add(product.getId());
+            }
+        }
+
+        for (Product overlappingProduct : overlappingProducts) {
+            boolean modifies = false;
+            Iterator<String> ppit = entitlementProducts.iterator();
             // No need to continue checking once we have found a modified product
             while (!modifies && ppit.hasNext()) {
-                modifies = p.modifies(ppit.next().getId());
+                modifies = overlappingProduct.modifies(ppit.next());
             }
             if (modifies) {
                 // Return all entitlements for the modified product
-                modifying.addAll(pidEnts.get(p.getId()));
+                modifying.addAll(pidEnts.getEntitlementsByProductId(overlappingProduct.getId()));
             }
         }
 
         return modifying;
     }
 
-    /*
-     * Add a productId to entitlement mapping, creating the collection if necessary
-     */
-    private void addProductIdToMap(Map<String, Set<Entitlement>> map,
-            String pid, Entitlement e) {
-        if (!map.containsKey(pid)) {
-            map.put(pid, new HashSet<Entitlement>());
-        }
-        map.get(pid).add(e);
+    public Set<Entitlement> listModifying(Entitlement entitlement) {
+        return batchListModifying(java.util.Arrays.asList(entitlement));
     }
 
-    /*
-     * Add an entitlement to the productId Entitlement map, using the entitlements
-     * productId as well as those if its provided products.
-     */
-    private void addToMap(Map<String, Set<Entitlement>> map, Entitlement e) {
-        addProductIdToMap(map, e.getPool().getProductId(), e);
-        for (Product pp : e.getPool().getProvidedProducts()) {
-            addProductIdToMap(map, pp.getId(), e);
-        }
+    public Set<Entitlement> listModifying(Collection entitlements) {
+        return batchListModifying(new ArrayList<Entitlement>(entitlements));
     }
 
+    public Map<Consumer, List<Entitlement>> getDistinctConsumers(
+            List<Entitlement> entsToRevoke) {
+        Map<Consumer, List<Entitlement>> result = new HashMap<Consumer, List<Entitlement>>();
+        for (Entitlement ent : entsToRevoke) {
+            if (!result.containsKey(ent.getConsumer())) {
```

Comment:
I think you can use the same `get key, if key null then create entry` idiom here that crog suggested elsewhere in the review.